src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -0,0 +1,254 @@
+
+            // Build the authorization request URI.
+            AuthorizationRequest request =
+                    new AuthorizationRequest.Builder(
+                                    new ResponseType(ResponseType.Value.CODE),
+                                    new ClientID(clientID))
+                            .scope(requestedScopes)
+                            .redirectionURI(redirectURI)
+                            .state(state)
+                            .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                            .endpointURI(authorizationEndpoint)
+                            .build();
+
+            // Open the browser to the authorization request URI.
+            if (Desktop.isDesktopSupported()) {
+                Desktop.getDesktop().browse(request.toURI());
+            } else {
+                logger.severe("Desktop operations not supported");
+                return null;
+            }
+
+            // Wait for the authorization response from the local HTTP server.
+            OidcResponse response = server.getOidcResponse(callbackContext.getTimeout());
+            if (response == null || !state.getValue().equals(response.getState())) {
+                logger.severe("OIDC response is null or returned an invalid state");
+                return null;
+            }
+
+            // Generate token request from the authorization code and PKCE verifier.
+            AuthorizationCode code = new AuthorizationCode(response.getCode());
+            AuthorizationCodeGrant codeGrant =
+                    new AuthorizationCodeGrant(code, redirectURI, codeVerifier);
+            TokenRequest tokenRequest =
+                    new TokenRequest(tokenEndpoint, new ClientID(clientID), codeGrant);
+
+            // Sends the token exchange request and parse the response to obtain tokens.
+            HTTPResponse httpResponse = tokenRequest.toHTTPRequest().send();
+            TokenResponse tokenResponse = OIDCTokenResponseParser.parse(httpResponse);
+            if (!tokenResponse.indicatesSuccess()) {
+                logger.severe("Token request failed with response: " + httpResponse.getBody());
+                return null;
+            }
+
+            return getOidcCallbackResultFromTokenResponse((OIDCTokenResponse) tokenResponse);
+        } catch (Exception e) {
+            logger.log(Level.SEVERE, "Error during OIDC authentication", e);
+            return null;
+        } finally {
+            try {
+                // Sleeping to ensure the server stays up long enough to respond to the browser's
+                // request after the redirect.
+                Thread.sleep(2000);
+            } catch (InterruptedException e) {
+                logger.log(Level.WARNING, "Thread interrupted", e);
+            }
+            server.stop();
+        }
+    }
+
+    public OidcCallbackResult doRefresh(OidcCallbackContext callbackContext) {
```

Comment:
This PR is mainly adding this `doRefresh()` function and the `TestOidcAuthFlowAndRefresh` class to test the refresh.